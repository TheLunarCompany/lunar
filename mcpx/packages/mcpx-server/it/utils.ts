import {
  accessLogFor,
  buildLogger,
  loggableError,
  LunarLogger,
} from "@mcpx/toolkit-core/logging";
import { Client } from "@modelcontextprotocol/sdk/client/index.js";
import { SSEClientTransport } from "@modelcontextprotocol/sdk/client/sse.js";
import { StreamableHTTPClientTransport } from "@modelcontextprotocol/sdk/client/streamableHttp.js";
import { Transport } from "@modelcontextprotocol/sdk/shared/transport.js";
import { MeterProvider } from "@opentelemetry/sdk-metrics";
import express from "express";
import { createServer, Server } from "http";
import { ConfigService } from "../src/config.js";
import { Config } from "../src/model/config/config.js";
import { TargetServer } from "../src/model/target-servers.js";
import { AuthGuard, noOpAuthGuard } from "../src/server/auth.js";
import { buildSSERouter } from "../src/server/sse.js";
import { buildStreamableHttpRouter } from "../src/server/streamable.js";
import { buildControlPlaneRouter } from "../src/server/control-plane.js";
import { Services } from "../src/services/services.js";
import {
  TESTKIT_SERVER_CALCULATOR,
  TESTKIT_SERVER_ECHO,
} from "../src/testkit/root.js";

const MCPX_PORT = 9000;

// Track all created loggers for cleanup
const allLoggers = new Set<LunarLogger>();

const getTestLogger: () => LunarLogger = () => {
  const logger = buildLogger({ logLevel: "debug", label: "test" });
  allLoggers.add(logger);
  return logger;
};

export const getMcpxLogger: () => LunarLogger = () => {
  const logger = buildLogger({ logLevel: "debug", label: "mcpx" });
  allLoggers.add(logger);
  return logger;
};

// Cleanup function to close all logger instances (for global teardown)
export const closeAllLoggers = (): void => {
  allLoggers.forEach((logger) => {
    try {
      logger.close();
    } catch (_e) {
      // Ignore errors when closing loggers
    }
  });
  allLoggers.clear();
};

const BASE_CONFIG: Config = {
  permissions: {
    default: { _type: "default-allow", block: [] },
    consumers: {},
  },
  toolGroups: [],
  auth: { enabled: false },
  toolExtensions: { services: {} },
};
export function buildConfig(props: Partial<Config> = {}): ConfigService {
  const { auth, permissions, toolGroups, toolExtensions } = {
    ...BASE_CONFIG,
    ...props,
  };
  return new ConfigService(
    { permissions, toolGroups, auth, toolExtensions },
    getMcpxLogger(),
  );
}

export const stdioTargetServers: TargetServer[] = [
  {
    type: "stdio",
    name: "echo-service",
    command: "node",
    args: [TESTKIT_SERVER_ECHO],
    env: {},
  },
  {
    type: "stdio",
    name: "calculator-service",
    command: "node",
    args: [TESTKIT_SERVER_CALCULATOR],
    env: {},
  },
];
export const oauthTargetServer: TargetServer = {
  type: "streamable-http",
  name: "oauth-mock-server",
  url: "http://localhost:9001/mcp",
};
export const allTargetServers: TargetServer[] = [
  ...stdioTargetServers,
  oauthTargetServer,
];
type TransportType = "SSE" | "StreamableHTTP";

export const transportTypes: TransportType[] = [
  "SSE",
  "StreamableHTTP",
] as const;

export class TestHarness {
  public clientConnectError?: Error | undefined;
  private loggers: LunarLogger[] = [];

  constructor(
    public client: Client,
    public server: Server,
    public services: Services,
    public testLogger: LunarLogger,
    private clientConnectExtraHeaders: Record<string, string> = {},
    private targetServers: TargetServer[] = stdioTargetServers,
  ) {
    // Track loggers for this harness instance
    this.loggers.push(testLogger);
  }

  addLogger(logger: LunarLogger): void {
    this.loggers.push(logger);
  }

  async initialize(transportType: TransportType): Promise<void> {
    // Setup MCPX
    await this.services.initialize();
    await Promise.all(
      this.targetServers.map((target) =>
        this.services.targetClients.addClient(target),
      ),
    );
    await this.server.listen(MCPX_PORT, () => {
      this.testLogger.info(`Test MCPX server listening on port ${MCPX_PORT}`);
    });

    const transport = this.buildTransport(transportType);

    await this.client
      .connect(transport)
      .then(() => this.testLogger.info("End-client connected"))
      .catch((e) => {
        const error = loggableError(e);
        this.testLogger.error("Failed to connect end-client", error);
        this.clientConnectError = e;
      });
  }

  async shutdown(): Promise<void> {
    await this.client.transport?.close();
    await this.client.close();
    await this.services.shutdown();
    await new Promise<void>((resolve) => {
      this.server.close(() => {
        this.testLogger.info("Test MCPX server closed");
        // Close loggers AFTER server has fully closed and all connection events have fired
        this.loggers.forEach((logger) => {
          try {
            logger.close();
          } catch (_e) {
            // Ignore errors when closing loggers
          }
        });
        this.loggers = [];
        resolve();
      });
    });
  }

  private buildTransport(transportType: TransportType): Transport {
    switch (transportType) {
      case "SSE":
        return new SSEClientTransport(
          new URL(`http://localhost:${MCPX_PORT}/sse`),
          {
            eventSourceInit: {
              fetch: (url, init) => {
                const headers = new Headers({
                  ...init?.headers,
                  ...this.clientConnectExtraHeaders,
                });
                return fetch(url, { ...init, headers });
              },
            },
          },
        );
      case "StreamableHTTP":
        return new StreamableHTTPClientTransport(
          new URL(`http://localhost:${MCPX_PORT}/mcp`),
          {
            requestInit: {
              headers: this.clientConnectExtraHeaders,
            },
          },
        );
      default:
        throw new Error(`Unknown transport type: ${transportType}`);
    }
  }
}

interface TestHarnessProps {
  config?: ConfigService;
  authGuard?: AuthGuard;
  mcpxLogger?: LunarLogger;
  clientConnectExtraHeaders?: Record<string, string>;
  targetServers?: TargetServer[];
}
function defaultTestHarnessProps(): Required<TestHarnessProps> {
  return {
    config: buildConfig(),
    authGuard: noOpAuthGuard,
    mcpxLogger: getMcpxLogger(),
    clientConnectExtraHeaders: {},
    targetServers: stdioTargetServers,
  };
}
export function getTestHarness(props: TestHarnessProps = {}): TestHarness {
  // the effective values are the defaults, unless overridden by props
  const {
    config,
    authGuard,
    mcpxLogger,
    clientConnectExtraHeaders,
    targetServers,
  } = {
    ...defaultTestHarnessProps(),
    ...props,
  };
  const testLogger = getTestLogger();
  const meterProvider = new MeterProvider();
  const services = new Services(config, meterProvider, testLogger);

  const client = new Client({ name: "end-client", version: "1.0.0" });

  // Wire router on server
  const sseRouter = buildSSERouter(authGuard, services, mcpxLogger);
  const streamableRouter = buildStreamableHttpRouter(
    authGuard,
    services,
    mcpxLogger,
  );
  const controlPlaneRouter = buildControlPlaneRouter(
    authGuard,
    services,
    mcpxLogger,
  );
  const app = express();
  const httpServer = createServer(app);
  app.use(express.json());
  app.use(accessLogFor(mcpxLogger));
  app.use(sseRouter);
  app.use(streamableRouter);
  app.use(controlPlaneRouter);

  const harness = new TestHarness(
    client,
    httpServer,
    services,
    testLogger,
    clientConnectExtraHeaders,
    targetServers,
  );

  // Track the mcpxLogger in this harness instance
  harness.addLogger(mcpxLogger);

  return harness;
}
