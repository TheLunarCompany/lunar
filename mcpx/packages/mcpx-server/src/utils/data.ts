// A function to remove `null`s and/or `undefined`s from an array
export function compact<X>(xs: Array<X | null | undefined>): X[] {
  return xs.flatMap((x) => (x ? [x] : []));
}

export function stringifyEq<T>(a: T, b: T): boolean {
  return JSON.stringify(a) === JSON.stringify(b);
}
