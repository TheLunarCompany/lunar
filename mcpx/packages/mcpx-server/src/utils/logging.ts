import { makeError } from "./errors.js";

export interface LoggableError {
  name: string;
  message: string;
  stack?: string;
}

export function loggableError(e: unknown): LoggableError {
  const error = makeError(e);
  return {
    name: error.name,
    message: error.message,
    stack: error.stack,
  };
}
