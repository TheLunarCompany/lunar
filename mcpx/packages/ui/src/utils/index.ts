export function createPageUrl(pageName: string) {
  return "/" + pageName.toLowerCase().replace(/ /g, "-");
}
