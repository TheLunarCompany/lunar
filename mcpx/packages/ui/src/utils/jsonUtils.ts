export type JsonValue =
  | string
  | number
  | boolean
  | null
  | undefined
  | JsonValue[]
  | { [key: string]: JsonValue };

export type JsonSchemaType = {
  type:
    | "string"
    | "number"
    | "integer"
    | "boolean"
    | "array"
    | "object"
    | "null";
  description?: string;
  required?: string[];
  default?: JsonValue;
  properties?: Record<string, JsonSchemaType>;
  items?: JsonSchemaType;
  enum?: JsonValue[];
};

export type JsonObject = { [key: string]: JsonValue };

export type DataType =
  | "string"
  | "number"
  | "bigint"
  | "boolean"
  | "symbol"
  | "undefined"
  | "object"
  | "function"
  | "array"
  | "null";

export function getDataType(value: JsonValue): DataType {
  if (Array.isArray(value)) return "array";
  if (value === null) return "null";
  return typeof value;
}

export function tryParseJson(str: string): {
  success: boolean;
  data: JsonValue;
} {
  const trimmed = str.trim();
  if (
    !(trimmed.startsWith("{") && trimmed.endsWith("}")) &&
    !(trimmed.startsWith("[") && trimmed.endsWith("]"))
  ) {
    return { success: false, data: str };
  }
  try {
    return { success: true, data: JSON.parse(str) };
  } catch {
    return { success: false, data: str };
  }
}

/**
 * Updates a value at a specific path in a nested JSON structure
 * @param obj The original JSON value
 * @param path Array of keys/indices representing the path to the value
 * @param value The new value to set
 * @returns A new JSON value with the updated path
 */
export function updateValueAtPath(
  obj: JsonValue,
  path: string[],
  value: JsonValue,
): JsonValue {
  if (path.length === 0) return value;

  if (obj === null || obj === undefined) {
    obj = !isNaN(Number(path[0])) ? [] : {};
  }

  if (Array.isArray(obj)) {
    return updateArray(obj, path, value);
  } else if (typeof obj === "object" && obj !== null) {
    return updateObject(obj as JsonObject, path, value);
  } else {
    console.error(
      `Cannot update path ${path.join(".")} in non-object/array value:`,
      obj,
    );
    return obj;
  }
}

/**
 * Updates an array at a specific path
 */
function updateArray(
  array: JsonValue[],
  path: string[],
  value: JsonValue,
): JsonValue[] {
  const [index, ...restPath] = path;
  const arrayIndex = Number(index);

  if (isNaN(arrayIndex)) {
    console.error(`Invalid array index: ${index}`);
    return array;
  }

  if (arrayIndex < 0) {
    console.error(`Array index out of bounds: ${arrayIndex} < 0`);
    return array;
  }

  let newArray: JsonValue[] = [];
  for (let i = 0; i < array.length; i++) {
    newArray[i] = i in array ? array[i] : null;
  }

  if (arrayIndex >= newArray.length) {
    const extendedArray: JsonValue[] = new Array(arrayIndex).fill(null);
    // Copy over the existing elements (now guaranteed to be dense)
    for (let i = 0; i < newArray.length; i++) {
      extendedArray[i] = newArray[i];
    }
    newArray = extendedArray;
  }

  if (restPath.length === 0) {
    newArray[arrayIndex] = value;
  } else {
    newArray[arrayIndex] = updateValueAtPath(
      newArray[arrayIndex],
      restPath,
      value,
    );
  }
  return newArray;
}

/**
 * Updates an object at a specific path
 */
function updateObject(
  obj: JsonObject,
  path: string[],
  value: JsonValue,
): JsonObject {
  const [key, ...restPath] = path;

  // Validate object key
  if (typeof key !== "string") {
    console.error(`Invalid object key: ${key}`);
    return obj;
  }

  const newObj = { ...obj };

  if (restPath.length === 0) {
    newObj[key] = value;
  } else {
    // Ensure key exists
    if (!(key in newObj)) {
      newObj[key] = {};
    }
    newObj[key] = updateValueAtPath(newObj[key], restPath, value);
  }
  return newObj;
}

/**
 * Gets a value at a specific path in a nested JSON structure
 * @param obj The JSON value to traverse
 * @param path Array of keys/indices representing the path to the value
 * @param defaultValue Value to return if path doesn't exist
 * @returns The value at the path, or defaultValue if not found
 */
export function getValueAtPath(
  obj: JsonValue,
  path: string[],
  defaultValue: JsonValue = null,
): JsonValue {
  if (path.length === 0) return obj;

  const [first, ...rest] = path;

  if (obj === null || obj === undefined) {
    return defaultValue;
  }

  if (Array.isArray(obj)) {
    const index = Number(first);
    if (isNaN(index) || index < 0 || index >= obj.length) {
      return defaultValue;
    }
    return getValueAtPath(obj[index], rest, defaultValue);
  }

  if (typeof obj === "object" && obj !== null) {
    if (!(first in obj)) {
      return defaultValue;
    }
    return getValueAtPath((obj as JsonObject)[first], rest, defaultValue);
  }

  return defaultValue;
}
