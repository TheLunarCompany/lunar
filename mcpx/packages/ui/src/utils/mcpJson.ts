import z from "zod/v4";

export const isValidJson = (value: string) => {
  try {
    JSON.parse(value);
    return true;
  } catch (e) {
    return false;
  }
};

export const serverNameSchema = z
  .string()
  .min(1, "Server name is required")
  .regex(
    /^[a-zA-Z0-9_-]+$/,
    "Server name can only contain letters, numbers, dashes (-), and underscores (_)"
  );

export const localServerSchema = z.strictObject({
  type: z.literal("stdio").default("stdio").optional(),
  command: z.string().min(1, "Command is required"),
  args: z.array(z.string()).default([]).optional(),
  env: z.record(z.string().min(1), z.string()).default({}).optional(),
  icon: z.string().optional(),
});

export const remoteServerSchema = z.strictObject({
  type: z.enum(["sse", "streamable-http"]).default("sse").optional(),
  url: z.url().and(
    z
      .string()
      .min(1, "URL is required")
      // Very simplified URL validation regex
      .regex(
        /^https?:\/\/[a-zA-Z0-9_-]+(\.[a-zA-Z0-9_\-]+)*(\:[0-9]+)?(\/[a-zA-Z0-9_\-]*)*$/,
      ),
  ),
  headers: z.record(z.string(), z.string()).optional(),
  icon: z.string().optional(),
});

export const mcpServerSchema = z.union([localServerSchema, remoteServerSchema]);

export const mcpJsonSchema = z.record(
  serverNameSchema,
  mcpServerSchema,
);

// The payload schemas are not strict, so they are used mainly to strip irrelevant properties,
// and do transforms that are not possible with the JSON Schemas, but not for validation.
export const localServerPayloadSchema = z
  .object({
    icon: z.string().optional(),
    name: z.string(),
    type: z.literal("stdio").optional(),
    command: z.string(),
    args: z
      .union([z.array(z.string()), z.string()])
      .default([])
      .optional(),
    env: z
      .union([z.record(z.string(), z.string()), z.string()])
      .default({})
      .optional(),
  })
  .transform((server) => ({
    ...server,
    type: server.type || "stdio",
    args: Array.isArray(server.args)
      ? server.args.join(" ")
      : server.args || "",
    env:
      typeof server.env === "string"
        ? server.env
        : JSON.stringify(server.env || {}),
  }));
export const remoteServerPayloadSchema = z.object({
  icon: z.string().optional(),
  name: z.string(),
  type: z.enum(["sse", "streamable-http"]).default("sse").optional(),
  url: z.string(),
  headers: z.record(z.string(), z.string()).optional(),
});

export const parseServerPayload = (server: z.input<typeof mcpServerSchema>) => {
  if (server.type === "stdio") {
    return localServerPayloadSchema.safeParse(server);
  }
  // Only set type if 'url' exists
  if ("url" in server && typeof server.url === "string") {
    const inferredType = inferServerTypeFromUrl(server.url);
    // Ensure type is always set and matches expected union
    if (inferredType === "sse") {
      return remoteServerPayloadSchema.safeParse({ ...server, type: "sse" });
    } else {
      return remoteServerPayloadSchema.safeParse({
        ...server,
        type: "streamable-http",
      });
    }
  }
  return remoteServerPayloadSchema.safeParse(server);
};

export const inferServerTypeFromUrl = (
  url: string,
): "sse" | "streamable-http" | undefined => {
  try {
    const urlObj = new URL(url);
    const pathname = urlObj.pathname.replace(/\/$/, "");
    if (pathname.endsWith("/sse")) {
      return "sse";
    } else if (pathname.endsWith("/mcp")) {
      return "streamable-http";
    }
    return undefined;
  } catch {
    return undefined;
  }
};

/**
 * Updates JSON content to include the inferred server type and icon
 * This ensures the saved configuration file includes the type field
 */
export const updateJsonWithServerType = (
  jsonContent: string,
  serverName: string,
  icon: string
): string => {
  try {
    const parsed = JSON.parse(jsonContent);
    const serverData = parsed[serverName];
    
    if (!serverData) {
      return jsonContent;
    }

    // Determine server type
    let serverType: string;
    if (serverData.type) {
      serverType = serverData.type;
    } else if ("url" in serverData) {
      serverType = inferServerTypeFromUrl(serverData.url) || "sse";
    } else {
      serverType = "stdio";
    }

    // Update the JSON with type and icon
    const updatedJson = {
      [serverName]: {
        ...serverData,
        type: serverType,
        icon: icon,
      }
    };

    return JSON.stringify(updatedJson, null, 2);
  } catch (error) {
    console.warn("Failed to update JSON with server type:", error);
    return jsonContent;
  }
};
