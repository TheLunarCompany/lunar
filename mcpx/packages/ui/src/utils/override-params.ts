import {
  ParamExtensionOverrideValue,
  ToolExtensionParamsRecord,
} from "@mcpx/shared-model";

const parseNumber = (v: unknown): number | undefined => {
  const n = typeof v === "number" ? v : Number(v);
  return !Number.isNaN(Number(v)) && Number.isFinite(n) ? n : undefined;
};
const parseBoolean = (v: unknown): boolean | undefined => {
  if (v === true || v === "true") return true;
  if (v === false || v === "false") return false;
  return undefined;
};
const parseValue = (type: ParamExtensionOverrideValue, raw: unknown) => {
  switch (type) {
    case "number":
      return parseNumber(raw);
    case "boolean":
      return parseBoolean(raw);
    case "string":
    case "array":
    case "object":
      return raw ?? undefined;
    default:
      return undefined;
  }
};

export const injectParamsListOverrides = (
  paramsList: { name: string; type: string; description: string }[],
  paramsValues: ToolExtensionParamsRecord,
) =>
  paramsList.map(({ name, type, description }) => ({
    name,
    type,
    description,
    value: parseValue(type, paramsValues[name]?.value),
  }));
