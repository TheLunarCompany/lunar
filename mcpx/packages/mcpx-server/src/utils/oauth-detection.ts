import { loggableError } from "@mcpx/toolkit-core/logging";
import { Logger } from "winston";
import z from "zod/v4";
import { RemoteTargetServer } from "../model/target-servers.js";

// These schemas are based on the OAuth 2.0 spec and conventions
const protectedResourceSchema = z.object({
  authorization_servers: z.array(z.unknown()).min(1),
});

const authServerSchema = z.object({
  authorization_endpoint: z.string().optional(),
  token_endpoint: z.string().optional(),
});

// Type for the fetch function to enable dependency injection
export type FetchFn = typeof fetch;

// Check if a remote server supports OAuth by looking for well-known endpoints
// Note that this implementation is based on the assumption that MCP servers will
// implement the OAuth spec correctly, or at least follow some of the conventions.
// This is a legitimate assumption since the MCP TS SDK uses this approach as well,
// to the best of our knowledge.
export async function detectOAuthSupport(
  targetServer: RemoteTargetServer,
  discoveryTimeoutMillis: number,
  logger: Logger,
  fetchFn: FetchFn = fetch,
): Promise<boolean> {
  try {
    const baseUrl = new URL(targetServer.url).origin;

    // Check both endpoints in parallel
    const [protectedResourceResult, authServerResult] =
      await Promise.allSettled([
        // Check protected resource metadata (correct per MCP spec, e.g. Notion)
        fetchFn(`${baseUrl}/.well-known/oauth-protected-resource`, {
          method: "GET",
          signal: AbortSignal.timeout(discoveryTimeoutMillis),
        }).then(async (response) => {
          if (!response.ok) return false;
          const json = await response.json();
          const parsed = protectedResourceSchema.safeParse(json);
          return parsed.success;
        }),

        // Check authorization server metadata (sometimes used by MCP servers, e.g. Linear)
        fetchFn(`${baseUrl}/.well-known/oauth-authorization-server`, {
          method: "GET",
          signal: AbortSignal.timeout(discoveryTimeoutMillis),
        }).then(async (response) => {
          if (!response.ok) return false;
          const json = await response.json();
          const parsed = authServerSchema.safeParse(json);
          return (
            parsed.success &&
            Boolean(
              parsed.data.authorization_endpoint || parsed.data.token_endpoint,
            )
          );
        }),
      ]);

    // If either result is `fulfilled` (i.e. successful HTTP call, including schema validation),
    // and valid, OAuth is assumed to be supported
    const protectedResourceSupported =
      protectedResourceResult.status === "fulfilled" &&
      protectedResourceResult.value;
    const authServerSupported =
      authServerResult.status === "fulfilled" && authServerResult.value;

    const oauthSupported = protectedResourceSupported || authServerSupported;

    logger.debug(
      `OAuth support ${oauthSupported ? "was" : "was not"} detected`,
      {
        name: targetServer.name,
        url: targetServer.url,
        basedOn: {
          protectedResource: protectedResourceSupported,
          authServer: authServerSupported,
        },
      },
    );
    return oauthSupported;
  } catch (error) {
    logger.debug("Error checking OAuth support", {
      name: targetServer.name,
      error: loggableError(error),
    });
    return false;
  }
}
