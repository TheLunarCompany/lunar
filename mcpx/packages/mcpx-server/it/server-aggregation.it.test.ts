import { getTestHarness, transportTypes } from "./utils.js";

describe.each(transportTypes)("%s Router", (transportType) => {
  const testHarness = getTestHarness();
  const client = testHarness.client;

  beforeAll(async () => {
    await testHarness.initialize(transportType);
  });

  afterAll(async () => {
    await testHarness.shutdown();
  });

  it("responds to listTools requests from end-client by listing target MCP server", async () => {
    const res = await client.listTools();
    expect(res.tools).toMatchSnapshot();
  });

  it("passes callTool requests from end-client to target MCP servers", async () => {
    const resEcho = await client.callTool({
      name: "echo-service__echo",
      arguments: { message: "The sound of silence?" },
    });
    expect(resEcho).toMatchSnapshot();

    const resAddition = await client.callTool({
      name: "calculator-service__add",
      arguments: { a: 5, b: 2 },
    });
    expect(resAddition).toMatchSnapshot();

    const resPowerOfTwo = await client.callTool({
      name: "calculator-service__powerOfTwo",
      arguments: { base: 4 },
    });
    expect(resPowerOfTwo).toMatchSnapshot();
  });
});
