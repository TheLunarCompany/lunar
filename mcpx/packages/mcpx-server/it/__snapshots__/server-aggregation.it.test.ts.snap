// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SSE Router passes callTool requests from end-client to target MCP servers 1`] = `
{
  "content": [
    {
      "text": "Tool echo: The sound of silence?",
      "type": "text",
    },
  ],
}
`;

exports[`SSE Router passes callTool requests from end-client to target MCP servers 2`] = `
{
  "content": [
    {
      "text": "Result: 7",
      "type": "text",
    },
  ],
}
`;

exports[`SSE Router passes callTool requests from end-client to target MCP servers 3`] = `
{
  "content": [
    {
      "text": "Result: 16",
      "type": "text",
    },
  ],
}
`;

exports[`SSE Router responds to listTools requests from end-client by listing target MCP server 1`] = `
[
  {
    "description": "Adds two numbers together",
    "inputSchema": {
      "$schema": "http://json-schema.org/draft-07/schema#",
      "additionalProperties": false,
      "properties": {
        "a": {
          "type": "number",
        },
        "b": {
          "type": "number",
        },
      },
      "required": [
        "a",
        "b",
      ],
      "type": "object",
    },
    "name": "calculator-service__add",
    "title": "Addition Tool",
  },
  {
    "description": "Calculates the power of two for a given number",
    "inputSchema": {
      "$schema": "http://json-schema.org/draft-07/schema#",
      "additionalProperties": false,
      "properties": {
        "base": {
          "type": "number",
        },
      },
      "required": [
        "base",
      ],
      "type": "object",
    },
    "name": "calculator-service__powerOfTwo",
    "title": "Power of Two Tool",
  },
  {
    "description": "Echoes back the provided message",
    "inputSchema": {
      "$schema": "http://json-schema.org/draft-07/schema#",
      "additionalProperties": false,
      "properties": {
        "message": {
          "type": "string",
        },
      },
      "required": [
        "message",
      ],
      "type": "object",
    },
    "name": "echo-service__echo",
    "title": "Echo Tool",
  },
]
`;

exports[`StreamableHTTP Router passes callTool requests from end-client to target MCP servers 1`] = `
{
  "content": [
    {
      "text": "Tool echo: The sound of silence?",
      "type": "text",
    },
  ],
}
`;

exports[`StreamableHTTP Router passes callTool requests from end-client to target MCP servers 2`] = `
{
  "content": [
    {
      "text": "Result: 7",
      "type": "text",
    },
  ],
}
`;

exports[`StreamableHTTP Router passes callTool requests from end-client to target MCP servers 3`] = `
{
  "content": [
    {
      "text": "Result: 16",
      "type": "text",
    },
  ],
}
`;

exports[`StreamableHTTP Router responds to listTools requests from end-client by listing target MCP server 1`] = `
[
  {
    "description": "Adds two numbers together",
    "inputSchema": {
      "$schema": "http://json-schema.org/draft-07/schema#",
      "additionalProperties": false,
      "properties": {
        "a": {
          "type": "number",
        },
        "b": {
          "type": "number",
        },
      },
      "required": [
        "a",
        "b",
      ],
      "type": "object",
    },
    "name": "calculator-service__add",
    "title": "Addition Tool",
  },
  {
    "description": "Calculates the power of two for a given number",
    "inputSchema": {
      "$schema": "http://json-schema.org/draft-07/schema#",
      "additionalProperties": false,
      "properties": {
        "base": {
          "type": "number",
        },
      },
      "required": [
        "base",
      ],
      "type": "object",
    },
    "name": "calculator-service__powerOfTwo",
    "title": "Power of Two Tool",
  },
  {
    "description": "Echoes back the provided message",
    "inputSchema": {
      "$schema": "http://json-schema.org/draft-07/schema#",
      "additionalProperties": false,
      "properties": {
        "message": {
          "type": "string",
        },
      },
      "required": [
        "message",
      ],
      "type": "object",
    },
    "name": "echo-service__echo",
    "title": "Echo Tool",
  },
]
`;
